### PR TITLE
Service can delete EC2 instances

### PR DESCRIPTION
**Pull request by amolgautam25**
Service can delete EC2 instance. 
Not the final commit, needs testing

### CODE REVIEW

File: pkg/awsController/ec2Manager.go [hunk 1]
```diff
@@ -0,0 +1,115 @@
+package awsController
```
Comment (fahlmant) [inline pkg/awsController/ec2Manager.go:1]: Can we please change the package name too? `awsManager`

File: pkg/awsManager/ec2Manager.go [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+	"github.com/aws/aws-sdk-go/service/ec2"
+	clientpkg "github.com/openshift/aws-account-shredder/pkg/aws"
+	"os"
+	"strings"
+)
+
+// function to list all the EC2 instances
+// not removing this function right now , as it is easy to verify after deletion if it was actually deleted or not
+// keeping for future testing purpose ( easy to use )
+func ListEc2Instances(assumedRoleClient clientpkg.Client) {
```
Comment (fahlmant) [inline pkg/awsManager/ec2Manager.go:16]: What's the functional difference between this and ListEc2InstancesForDeletion? Both do the same thing, not sure we need them both

File: pkg/awsManager/ec2Manager.go [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+	token := ""
+	for {
+		ec2Descriptions, err := assumedRoleClient.DescribeInstances(&ec2.DescribeInstancesInput{NextToken: aws.String(token)})
+		if err != nil {
+			fmt.Println("ERROR:", err)
+			os.Exit(1)
+		}
+
+		// nested for loop to read the tags , as it is a part of structure inside a structure output. Refer : https://pkg.go.dev/github.com/aws/aws-sdk-go/service/ec2?tab=doc#DescribeInstancesOutput
+		for _, iterator := range ec2Descriptions.Reservations {
```
Comment (fahlmant) [inline pkg/awsManager/ec2Manager.go:49]: `for _, reservation := ...`

File: pkg/awsManager/ec2Manager.go [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+			for _, iterator2 := range iterator.Instances {
```
Comment (fahlmant) [inline pkg/awsManager/ec2Manager.go:50]: `for _, instance := ...`

File: pkg/awsManager/ec2Manager.go [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+				for _, iterator3 := range iterator2.Tags {
```
Comment (fahlmant) [inline pkg/awsManager/ec2Manager.go:51]: Same as the above

File: pkg/awsManager/ec2Manager.go [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+
+					//checking if the tags start with 'kubernetes.io' and status code !=48
+					if strings.HasPrefix(*iterator3.Key, "kubernetes.io") && (*iterator2.State.Code != 48) {
+						// some debugging statements , will be removed after thorough testing.
```
Comment (fahlmant) [inline pkg/awsManager/ec2Manager.go:55]: Any reason to keep all these comments in here?

File: pkg/awsManager/ec2Manager.go [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+			default:
+				fmt.Println(err.Error())
+			}
+		} else {
+			fmt.Println(err.Error())
+		}
+	}
+
+	// for debugging , will be removed afterwards
+	fmt.Println(result)
```
Comment (fahlmant) [inline pkg/awsManager/ec2Manager.go:113]: Can we remove this?

File: pkg/awsManager/ec2Manager.go [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+
+}
+
+func CleanEc2Instacnes(assumedRoleClient clientpkg.Client) {
+
+	EC2InstancesToBeDeleted := ListEc2InstancesForDeletion(assumedRoleClient)
```
Comment (fahlmant) [inline pkg/awsManager/ec2Manager.go:119]: `ec2InstancesToBeDeleted` not a big deal, but in Go any var that starts with a capital letter is exported. We should avoid that when possible.

Comment (amolgautam25) [reply pkg/awsManager/ec2Manager.go:119]: resolved.

File: pkg/awsManager/ec2Manager.go (hunk 3 above)
Comment (amolgautam25) [reply pkg/awsManager/ec2Manager.go]: i have removed the func ListEc2Instances()

File: pkg/awsManager/ec2Manager.go [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+		}
+	}
+
+	return EC2InstancesToBeDeleted
+}
+
+// this takes all the instances from insancesToBeDeleted  ( one by one ) and deletes them
+func DeleteEc2Instance(client clientpkg.Client, EC2InstancesToBeDeleted []*string) {
+
+	if EC2InstancesToBeDeleted == nil {
```
Comment (fahlmant) [inline pkg/awsManager/ec2Manager.go:62]: Can you test that this will be hit?

Comment (amolgautam25) [reply pkg/awsManager/ec2Manager.go:62]: Confirmed , it hits. 
" EC2 instances in region  ap-southeast-2
inside the if condition . it Hits 

 EC2 instances in region  sa-east-1
inside the if condition . it Hits 
"

output ommited for other regions.

File: pkg/awsManager/ec2Manager.go [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+// this only creates an array of pointers and does not delete the instances
+func ListEc2InstancesForDeletion(client clientpkg.Client) []*string {
+
+	var EC2InstancesToBeDeleted []*string
+	token := ""
+	for {
+		ec2Descriptions, err := client.DescribeInstances(&ec2.DescribeInstancesInput{NextToken: aws.String(token)})
+		if err != nil {
+			fmt.Println("ERROR:", err)
+			os.Exit(1)
```
Comment (fahlmant) [inline pkg/awsManager/ec2Manager.go:23]: Please remove this

Comment (amolgautam25) [reply pkg/awsManager/ec2Manager.go]: removed ,and removed the same in main,go as well

**Pull request merged**